### DISCUSSION TITLE
Share Your Shell 'Plugins'

### DISCUSSION THREAD

**Discussion by cmacrae** [cmacrae/spacebar#70] [Show and tell]
# 🐚 
Got something cool working with spacebar's `shell` options? 😎 Share the config and/or script!  
Want to implement something but you're not quite sure how? 🤔 Go ahead and ask!

_Note: Providing more of these "slots" is planned for a future release_

**Comment by cmacrae** [comment]
I'll get things kicked off with a couple simple outputs I've been using :)

### Unread e-mail indicator
![image](https://user-images.githubusercontent.com/3392199/114210135-514ae780-9957-11eb-90d3-c63a4c4e4e05.png)
I use [mu/mu4e](https://djcbsoftware.nl/code/mu/) for e-mail inside Emacs. I can lose track of my e-mail throughout the day, but have been doing better since I added this (apart from today, clearly!)
```
spacebar -m config right_shell on
spacebar -m config right_shell_command "mu find 'm:/work/inbox' flag:unread | wc -l | tr -d \"[:blank:]\""
spacebar -m config right_shell_icon 
```
As you can see, it's for my work inbox, but could probably quite easily be switched out to evaluate `mu4e-context-current` via `emacsclient`

### Current Kubernetes context/namespace
![image](https://user-images.githubusercontent.com/3392199/114214797-b35a1b80-995c-11eb-80ef-34f4b606bbaf.png)
One for the engineers :)
```
spacebar -m config right_shell_command /path/to/script.sh
spacebar -m config right_shell_icon 
```
Where `script.sh` is:
```bash
#! /usr/bin/env bash

CONTEXT=$(kubectl config current-context)
NAMESPACE=$(kubectl config view --minify --output 'jsonpath={..namespace}')
echo "${CONTEXT}:${NAMESPACE}"

```

**Comment by johnallen3d** [comment]
### Current track information from `mpd`

<img width="553" alt="CleanShot 2021-04-09 at 20 25 42@2x" src="https://user-images.githubusercontent.com/184915/114252142-ccb99280-9971-11eb-84b5-011d1b46cb31.png">

Showing just artist, album and status for `mpd` using `mpc`.

#### Configuration

```sh
spacebar -m config title                off
spacebar -m config center_shell         on
spacebar -m config center_shell_command "$HOME/.config/spacebar/mpd"
```

#### Script

```sh
#! /usr/bin/env bash

STATUS="$(/usr/local/bin/mpc status 2>/dev/null)"

if [ $(echo "$STATUS" | wc -l | tr -d ' ') == "1" ]; then
  output="Queue up some tunes!"
else
  artist=$(echo "$STATUS" | awk -F" - " 'NR==1{print $1}')
  song=$(echo "$STATUS"   | awk -F" - " 'NR==1{print $2}')
  status=$(echo "$STATUS" | awk         'NR==2{print $1}')

  if [ $status = "[playing]" ]; then
    status="Playing"
  else
    status="Paused"
  fi

  output="${artist} • ${song} • ${status}"
fi

echo $output
```

**Comment by muhammad-saleh** [comment]
CPU Utilization:

<img width="410" alt="image" src="https://user-images.githubusercontent.com/1510608/115126769-f0b05000-9fd1-11eb-9249-9188f6d47184.png">



Config:
```
spacebar -m config right_shell_command  "$HOME/spacebar_scripts/cpu.sh"
```


Script:
```
#! /usr/bin/env bash

ps -axro pcpu | awk '{sum+=$1} END {print "%"sum}'
```

**Comment by ericlovesmath** [comment]
Keyboard Language:

![image](https://user-images.githubusercontent.com/17755970/116189668-27771c00-a6de-11eb-8fb4-f49b419c201c.png)
![image](https://user-images.githubusercontent.com/17755970/116189638-1b8b5a00-a6de-11eb-9e85-70fd0176648a.png)

Config:

`spacebar -m config right_shell_command ~/bin/keyboard-layout.sh
`

Script:

```
#!/usr/bin/env bash

osascript -e 'tell application "System Events" to tell process "SystemUIServer" to get the value of the first menu bar item of menu bar 1 whose description is "text input"'
```
That being said, it would be helpful to have a "force spacebar to update" function as there is a severe delay, the shell script running only every 5 seconds

**Reply by cmacrae** [reply]
Very nice! [As pointed out in the issue you raised](https://github.com/cmacrae/spacebar/issues/76#issuecomment-827446207), you can expect higher refresh frequencies (per module) in the rewrite :) You can also hack around this limitation in the current implementation with:  
```
while sleep 1 ; do spacebar -m config left_shell_command "date" ; done
```

For example. This will draw the shell output of `date` every second

**Reply by basbebe** [reply]
> ```bash
> osascript -e 'tell application "System Events" to tell process "SystemUIServer" to get the value of the first menu bar item of menu bar 1 whose description is "text input"'
> ```

This does not work for me (macOS Big Sur).

These work for me, however ([source](https://stackoverflow.com/a/21613702/1086156))
```bash
defaults read ~/Library/Preferences/com.apple.HIToolbox.plist AppleSelectedInputSources | egrep -w'KeyboardLayout Name' | sed -E 's/^.+ = \"?([^\"]+)\"?;$/\\1/'
```
```bash
defaults read ~/Library/Preferences/com.apple.HIToolbox.plist AppleSelectedInputSources | command rg -e '"KeyboardLayout Name" = "*([^"]*)"*;' --replace '$1' --only-matching --color never
```

**Comment by anujc4** [comment]
Wifi SSID and speed

network.sh
```
#!/bin/bash

output=$(/System/Library/PrivateFrameworks/Apple80211.framework/Versions/A/Resources/airport -I)
airport=$(echo "$output" | grep 'AirPort' | awk -F': ' '{print $2}')

if [ "$airport" = "Off" ]; then
  echo "Offline"
else
  echo "$(echo "$output" | grep ' SSID' | xargs | awk -F': ' 'echo $2')" \
    "$(echo "$output" | grep 'lastTxRate' | awk -F': ' '{print $2}')" Mb/s
fi
```

spacebarrc
```
spacebar -m config right_shell        on
spacebar -m config right_shell_icon   
spacebar -m config right_shell_command "$HOME/.config/spacebar/network_speed.sh"
```

![image](https://user-images.githubusercontent.com/9458914/116228174-d6ead780-a74c-11eb-9a0a-501feaf66375.png)

**Reply by anujc4** [reply]
Additionally, the "SSID: " prefix can also be removed by appending ` | sed 's/SSID: //')` to the output

**Reply by cmacrae** [reply]
Very nice! 👏

**Comment by andynameistaken** [comment]
I wrote  script for checking Wi-Fi signal quality (RSSI / Noise) : 

```
#!/usr/bin/env bash

# Created by Andrzej Zahorski :)
result=$(/System/Library/PrivateFrameworks/Apple80211.framework/Versions/A/Resources/airport -I \
    | grep "CtlRSSI\|agrCtlNoise" \
    | sed -e 's/^.*://g'  | { read -r a; read -r b; printf '%s\n' $((a-b)); })

printf 'Wi-Fi: '
# echo $result
if ((result > 40))
then
    echo "[++++]"
elif ((result > 25 && result <= 40))
then
    echo "[+++-]"
elif ((result > 15 && result <= 25))
then
    echo "[++--]"
elif ((result > 10 && result <= 15))
then
    echo "[+---]"
elif (( result > 0 &&result < 10  ))
then
    echo "[----]"
elif (( result == 0 ))
then
    echo "Off"
fi

```
![Screenshot 2021-05-09 at 17 30 40](https://user-images.githubusercontent.com/15692002/117577950-56cd5600-b0ec-11eb-8d09-43a632fc0490.png)

**Comment by joellidin** [comment]
I wrote a quick python script for collecting the current bitcoin price and the 24 hours percentage change from Coinmarketcap. However, it requires that you python3 installation has bs4 and requests.

```config
spacebar -m config right_shell          on
spacebar -m config right_shell_icon     ""
spacebar -m config right_shell_command  "$HOME/.scripts/crypto"
```

```python
#!/usr/bin/env python3
from bs4 import BeautifulSoup
import requests
import json
import sys

fiat = 'BTC'
if len(sys.argv) > 1:
    fiat = sys.argv[1]

if fiat.lower() == 'eth':
    cur = 1
    idx = '1027'
else:
    cur = 0
    idx = '1'
url = 'https://coinmarketcap.com/currencies/bitcoin'

cmc = requests.get(url)
soup = BeautifulSoup(cmc.content, 'html.parser')
data = soup.find('script', id='__NEXT_DATA__', type='application/json')
data = json.loads(data.contents[0])
crypt_data = data['props']['initialState']['cryptocurrency']['quotes'][idx]

percentChange = crypt_data['fmc24hpc']
price = crypt_data['p']
# percentChangeFormat = '<span color="{}">{}{:.2f}%</span>'
percentChangeFormat = '{}{:.2f}%'
precision = 2
if percentChange > 0:
    # percentChangeInfo = percentChangeFormat.format('#3BB92D', '', percentChange)
    percentChangeInfo = percentChangeFormat.format('', percentChange)
elif percentChange == 0:
    # percentChangeInfo = percentChangeFormat.format('#CCCCCC', '', percentChange)
    percentChangeInfo = percentChangeFormat.format('', percentChange)
else:
    # percentChangeInfo = percentChangeFormat.format('#F7555E', '', percentChange)
    percentChangeInfo = percentChangeFormat.format('', -percentChange)
print(('${:.' + str(precision) + 'f} {}').format(price, percentChangeInfo))  # Full Text
```
<img width="371" alt="Screenshot 2021-05-13 at 14 43 35" src="https://user-images.githubusercontent.com/19657866/118127140-9bf9cc80-b3f9-11eb-96b1-66cf5b539639.png">

**Comment by Ducky6944** [comment]
Here's a script to display the Current State of the Music App and/or current track information. It could use some tweaking particularly for long titles. It's kind of sloppy, but it works pretty well. 

If anyone knows if there's a way to dynamically update the icon that would be great. ~~My full script switches between a few different things like, stocks, music, etc.~~ I've tried to put it in the script but it doesn't seem to work and spacebar stops responding. 

```OSASCRIPT
MUSIC_INFO=$(osascript <<EOF
on titleCase(theString)
	set UpperFirstCharString to do shell script "echo " & character 1 of theString & " | tr [:lower:] [:upper:]"
	set theString to UpperFirstCharString & characters 2 through -1 of theString
	return theString
end titleCase

on appRunning(appName)
	if application appName is running then
		return true
	else
		return false
	end if
end appRunning

on musicPlaying()
	tell application "Music"
		set playerState to (get player state) as text
	end tell
	
	if playerState is "paused" then
		set stateIcon to " "
	else if playerState is "stopped" then
		set stateIcon to " "
	else if playerState is "playing" then
		set stateIcon to " "
	end if
		
	if playerState is not "stopped" then
		return {true, titleCase(playerState), stateIcon}
	else
		return {false, titleCase(playerState), stateIcon}
	end if
end musicPlaying

on getTrackInfo()
	set appRunning to appRunning("Music")
	set musicPlaying to musicPlaying()	
	
	if appRunning is true then
		if item 1 of musicPlaying is true then
			tell application "Music"
				set info to get {name, artist, album} of current track
				return "" & item 3 of musicPlaying & "  " & item 1 of info & " x " & item 2 of info & " - " & item 3 of info
			end tell
		else
			return item 3 of musicPlaying & "  Music Stopped "
		end if
	else
		return "    Music App Not Running "
	end if
end getTrackInfo

getTrackInfo()
EOF
)

printf "$MUSIC_INFO" 
```

**Reply by Meaticus22** [reply]
Used the same idea but made changes so it applies to spotify.app 
 ```
MUSIC_INFO=$(osascript <<EOF
on titleCase(theString)
	set UpperFirstCharString to do shell script "echo " & character 1 of theString & " | tr [:lower:] [:upper:]"
	set theString to UpperFirstCharString & characters 2 through -1 of theString
	return theString
end titleCase

on appRunning(appName)
	if application appName is running then
		return true
	else
		return false
	end if
end appRunning

on getTrackInfo()
	set appRunning to appRunning("Spotify")
	
	if appRunning is true then
		tell application "Spotify"
			set info to get {name, artist} of current track
			return "  " & item 1 of info & " x " & item 2 of info
		end tell
	else
		return "    Spotify App Not Running "
	end if
end getTrackInfo

getTrackInfo()
EOF
)

printf "$MUSIC_INFO"
```
<img width="1919" alt="image" src="https://github.com/cmacrae/spacebar/assets/117058929/24d94e84-713d-4cab-86bf-5eaf840fd839">

```
spacebar -m config left_shell on
spacebar -m config left_shell_command "sh $HOME/.config/spacebar/spotify.sh"
spacebar -m config left_shell_icon 
spacebar -m config left_shell_icon_color 0xfff5f0f0
```
